File: frontend/src/utils/shortcuts.js
import { formatShortcutLabel, isMacPlatform } from './platform'
import {
  expandSendString,
  matchesKeyMapBinding,
} from './keymaps'
import { resolveSnippetCommand } from './snippetVariables'

/** 默认快捷键（useMod=true 表示 Cmd/Ctrl） */
export const DEFAULT_SHORTCUTS = {
  newWindow: { key: 'n', useMod: true },
  machineConfig: { key: 'm', useMod: true },
  connectionManager: { key: 'e', useMod: true },
  envVars: { key: 'u', useMod: true },
  systemSettings: { key: ',', useMod: true },
  refreshConfig: { key: 'r', useMod: true },
  find: { key: 'f', useMod: true },
  copy: { key: 'c', useMod: true },
  paste: { key: 'v', useMod: true },
  clearOutput: { key: 'k', useMod: true },
  commandPalette: { key: 'p', useMod: true, useShift: true },
  paneZoom: { key: 'z', useMod: true, useShift: true },
  nextTab: { key: 'Tab', useMod: true },
  prevTab: { key: 'Tab', useMod: true, useShift: true },
  closeTab: { key: 'w', useMod: true },
  toggleBroadcast: { key: 'b', useMod: true },
  openSftp: { key: 'o', useMod: true, useShift: true },
  openLocalShell: { key: 'l', useMod: true },
  splitFocusLeft: { key: 'ArrowLeft', useMod: true, useAlt: true },
  splitFocusRight: { key: 'ArrowRight', useMod: true, useAlt: true },
  splitFocusUp: { key: 'ArrowUp', useMod: true, useAlt: true },
  splitFocusDown: { key: 'ArrowDown', useMod: true, useAlt: true },
}

export const SHORTCUT_LABELS = {
  newWindow: '新建窗口',
  machineConfig: '机器配置',
  connectionManager: '连接 / 快速切换',
  envVars: '环境变量',
  systemSettings: '系统设置',
  refreshConfig: '刷新配置列表',
  find: '查找',
  copy: '复制',
  paste: '粘贴',
  clearOutput: '清空输出',
  commandPalette: '命令面板（历史/片段）',
  paneZoom: '分屏窗格最大化',
  nextTab: '下一个标签',
  prevTab: '上一个标签',
  closeTab: '关闭当前标签',
  toggleBroadcast: '开关命令广播',
  openSftp: '打开/折叠文件面板',
  openLocalShell: '打开本机终端',
  splitFocusLeft: '分屏焦点←',
  splitFocusRight: '分屏焦点→',
  splitFocusUp: '分屏焦点↑',
  splitFocusDown: '分屏焦点↓',
}

/** 设置页分组展示（覆盖全部可配置项） */
export const SHORTCUT_GROUPS = [
  {
    title: '应用',
    ids: ['newWindow', 'systemSettings', 'machineConfig', 'connectionManager', 'envVars', 'openLocalShell'],
  },
  {
    title: '编辑与输出',
    ids: ['find', 'copy', 'paste', 'clearOutput', 'refreshConfig'],
  },
  {
    title: 'Shell 标签与分屏',
    ids: [
      'commandPalette',
      'paneZoom',
      'nextTab',
      'prevTab',
      'closeTab',
      'toggleBroadcast',
      'openSftp',
      'splitFocusLeft',
      'splitFocusRight',
      'splitFocusUp',
      'splitFocusDown',
    ],
  },
]

export function mergeShortcuts(partial) {
  const result = {}
  for (const [id, def] of Object.entries(DEFAULT_SHORTCUTS)) {
    const cur = partial?.[id]
    result[id] = {
      key: (cur?.key != null && String(cur.key).length > 0) ? String(cur.key) : def.key,
      useMod: cur?.useMod !== undefined ? !!cur.useMod : !!def.useMod,
      useShift: cur?.useShift !== undefined ? !!cur.useShift : !!def.useShift,
      useAlt: cur?.useAlt !== undefined ? !!cur.useAlt : !!def.useAlt,
    }
  }
  return result
}

/** 判断键盘事件是否匹配绑定 */
export function matchesShortcut(e, binding) {
  if (!binding || !binding.key || !e) return false
  const needMod = binding.useMod !== false
  const hasMod = !!(e.metaKey || e.ctrlKey)
  if (needMod !== hasMod) return false

  const needAlt = !!binding.useAlt
  if (needAlt !== !!e.altKey) return false

  const needShift = !!binding.useShift
  if (needShift !== !!e.shiftKey) return false

  const target = String(binding.key)
  if (target === ',') {
    return e.key === ',' || e.code === 'Comma'
  }
  if (target === 'Tab') {
    return e.key === 'Tab' || e.code === 'Tab'
  }
  if (target.startsWith('Arrow')) {
    return e.key === target || e.code === target
  }

  const pressed = e.key === 'Escape' ? 'Escape' : String(e.key || '')
  if (pressed.toLowerCase() === target.toLowerCase()) return true

  const code = String(e.code || '')
  if (/^Key[A-Z]$/i.test(code)) {
    return code.slice(3).toLowerCase() === target.toLowerCase()
  }
  if (/^Digit[0-9]$/.test(code)) {
    return code.slice(5) === target
  }
  return false
}

/**
 * 检测快捷键表内部冲突。返回 [{ a, b, label }]
 */
export function findShortcutConflicts(shortcuts) {
  const entries = Object.entries(shortcuts || {}).filter(([, b]) => b?.key)
  const conflicts = []
  for (let i = 0; i < entries.length; i++) {
    for (let j = i + 1; j < entries.length; j++) {
      const [idA, a] = entries[i]
      const [idB, b] = entries[j]
      if (
        String(a.key).toLowerCase() === String(b.key).toLowerCase()
        && !!a.useMod === !!b.useMod
        && !!a.useShift === !!b.useShift
        && !!a.useAlt === !!b.useAlt
      ) {
        conflicts.push({
          a: idA,
          b: idB,
          label: `${SHORTCUT_LABELS[idA] || idA} 与 ${SHORTCUT_LABELS[idB] || idB}`,
        })
      }
    }
  }
  return conflicts
}

export function formatShortcut(binding, isMac = isMacPlatform()) {
  return formatShortcutLabel(binding, isMac)
}

/** 是否为 xterm 内部输入框 */
export function isXtermInput(el) {
  if (!el || el.nodeType !== 1) return false
  return !!el.classList?.contains('xterm-helper-textarea')
}

/**
 * 是否为应跳过全局快捷键的表单输入（不含 xterm）。
 * 终端聚焦时焦点在 textarea 上，不能当作普通输入框屏蔽。
 */
export function isFormFieldTarget(el) {
  if (!el || el.nodeType !== 1) return false
  if (isXtermInput(el)) return false
  const tag = el.tagName
  if (tag === 'INPUT' || tag === 'TEXTAREA' || tag === 'SELECT') return true
  if (el.isContentEditable || el.contentEditable === 'true') return true
  return false
}

/** 从 KeyboardEvent 生成绑定（用于设置页录制） */
export function bindingFromEvent(e) {
  let key = e.key === 'Escape' ? '' : e.key
  if (!key || key === 'Control' || key === 'Meta' || key === 'Alt' || key === 'Shift') {
    return null
  }
  if (key === 'Tab' || key.startsWith('Arrow')) {
    // keep
  } else if (key.length === 1) {
    key = key.toLowerCase()
  }
  return {
    key: key === ',' ? ',' : key,
    useMod: true,
    useShift: !!e.shiftKey,
    useAlt: !!e.altKey,
  }
}

export function emptySnippetBinding() {
  return { key: '', useMod: false, useAlt: false, useShift: false }
}

/** 规范化单条命令片段 */
export function normalizeSnippet(raw, index = 0) {
  const binding = raw?.binding
  return {
    id: raw?.id || `sn-${Date.now()}-${index}`,
    name: raw?.name != null ? String(raw.name) : '',
    command: raw?.command != null ? String(raw.command) : '',
    scope: raw?.scope ? String(raw.scope) : 'global',
    execute: raw?.execute !== undefined ? !!raw.execute : true,
    onConnect: !!raw?.onConnect,
    binding: {
      key: binding?.key != null ? String(binding.key) : '',
      useMod: !!binding?.useMod,
      useAlt: !!binding?.useAlt,
      useShift: !!binding?.useShift,
    },
  }
}

export function normalizeSnippets(list) {
  return (Array.isArray(list) ? list : []).map((s, i) => normalizeSnippet(s, i))
}

/** Shell 终端快捷键：匹配绑定了组合键的片段 */
export function findMatchingSnippet(e, snippets) {
  if (!e || !Array.isArray(snippets)) return null
  for (const s of snippets) {
    if (!s?.binding?.key) continue
    if (matchesKeyMapBinding(e, s.binding)) return s
  }
  return null
}

/** 片段发送内容：支持转义；execute 时自动补换行 */
export async function buildSnippetPayload(snippet, { promptVars = true } = {}) {
  const resolved = await resolveSnippetCommand(snippet, { prompt: promptVars })
  if (resolved == null) return ''
  let text = expandSendString(resolved)
  if (snippet?.execute && text && !/[\r\n]$/.test(text)) {
    text += '\n'
  }
  return text
}

File: data/shortcuts.go
package data

import (
	"fmt"
	"strings"
)

// ShortcutBinding 单条快捷键绑定
type ShortcutBinding struct {
	Key      string `json:"key"`
	UseMod   bool   `json:"useMod"`
	UseShift bool   `json:"useShift,omitempty"`
	UseAlt   bool   `json:"useAlt,omitempty"`
}

// ShellSnippet 终端命令片段（可绑定快捷键；可选择是否直接执行）
type ShellSnippet struct {
	ID      string         `json:"id"`
	Name    string         `json:"name"`
	Command string         `json:"command"`
	Scope   string         `json:"scope,omitempty"` // global 或机器配置名
	Binding *KeyMapBinding `json:"binding,omitempty"`
	// Execute 为 true 时发送到终端后追加换行并执行；false 仅插入文本
	Execute bool `json:"execute"`
	// OnConnect 为 true 时，Shell 会话连接成功后自动插入/执行（作用域为 global 或匹配机器）
	OnConnect bool `json:"onConnect,omitempty"`
}

// ShortcutSettings 可自定义系统快捷键（独立 JSON 文件）
type ShortcutSettings struct {
	NewWindow         ShortcutBinding `json:"newWindow"`
	MachineConfig     ShortcutBinding `json:"machineConfig"`
	ConnectionManager ShortcutBinding `json:"connectionManager"`
	EnvVars           ShortcutBinding `json:"envVars"`
	SystemSettings    ShortcutBinding `json:"systemSettings"`
	RefreshConfig     ShortcutBinding `json:"refreshConfig"`
	Find              ShortcutBinding `json:"find"`
	Copy              ShortcutBinding `json:"copy"`
	Paste             ShortcutBinding `json:"paste"`
	ClearOutput       ShortcutBinding `json:"clearOutput"`
	CommandPalette    ShortcutBinding `json:"commandPalette"`
	// QuickSwitcher 已废弃（合并进 ConnectionManager）；保留字段以兼容旧配置
	QuickSwitcher     ShortcutBinding `json:"quickSwitcher,omitempty"`
	PaneZoom          ShortcutBinding `json:"paneZoom"`
	NextTab           ShortcutBinding `json:"nextTab"`
	PrevTab           ShortcutBinding `json:"prevTab"`
	CloseTab          ShortcutBinding `json:"closeTab"`
	ToggleBroadcast   ShortcutBinding `json:"toggleBroadcast"`
	OpenSftp          ShortcutBinding `json:"openSftp"`
	OpenLocalShell    ShortcutBinding `json:"openLocalShell"`
	SplitFocusLeft    ShortcutBinding `json:"splitFocusLeft"`
	SplitFocusRight   ShortcutBinding `json:"splitFocusRight"`
	SplitFocusUp      ShortcutBinding `json:"splitFocusUp"`
	SplitFocusDown    ShortcutBinding `json:"splitFocusDown"`
	Snippets          []ShellSnippet  `json:"snippets,omitempty"`
}

// DefaultShortcutSettings 默认快捷键
func DefaultShortcutSettings() ShortcutSettings {
	return ShortcutSettings{
		NewWindow:         ShortcutBinding{Key: "n", UseMod: true},
		MachineConfig:     ShortcutBinding{Key: "m", UseMod: true},
		ConnectionManager: ShortcutBinding{Key: "e", UseMod: true},
		EnvVars:           ShortcutBinding{Key: "u", UseMod: true},
		SystemSettings:    ShortcutBinding{Key: ",", UseMod: true},
		RefreshConfig:     ShortcutBinding{Key: "r", UseMod: true},
		Find:              ShortcutBinding{Key: "f", UseMod: true},
		Copy:              ShortcutBinding{Key: "c", UseMod: true},
		Paste:             ShortcutBinding{Key: "v", UseMod: true},
		ClearOutput:       ShortcutBinding{Key: "k", UseMod: true},
		CommandPalette:    ShortcutBinding{Key: "p", UseMod: true, UseShift: true},
		PaneZoom:          ShortcutBinding{Key: "z", UseMod: true, UseShift: true},
		NextTab:           ShortcutBinding{Key: "Tab", UseMod: true},
		PrevTab:           ShortcutBinding{Key: "Tab", UseMod: true, UseShift: true},
		CloseTab:          ShortcutBinding{Key: "w", UseMod: true},
		ToggleBroadcast:   ShortcutBinding{Key: "b", UseMod: true},
		OpenSftp:          ShortcutBinding{Key: "o", UseMod: true, UseShift: true},
		OpenLocalShell:    ShortcutBinding{Key: "l", UseMod: true},
		SplitFocusLeft:    ShortcutBinding{Key: "ArrowLeft", UseMod: true, UseAlt: true},
		SplitFocusRight:   ShortcutBinding{Key: "ArrowRight", UseMod: true, UseAlt: true},
		SplitFocusUp:      ShortcutBinding{Key: "ArrowUp", UseMod: true, UseAlt: true},
		SplitFocusDown:    ShortcutBinding{Key: "ArrowDown", UseMod: true, UseAlt: true},
	}
}

func fillShortcutDefaults(s *ShortcutSettings) {
	def := DefaultShortcutSettings()
	if s.NewWindow.Key == "" {
		s.NewWindow = def.NewWindow
	}
	if s.MachineConfig.Key == "" {
		s.MachineConfig = def.MachineConfig
	}
	// 旧版默认 Mod+E 绑定环境变量；升级后改为连接管理器，环境变量迁到 Mod+U
	if s.ConnectionManager.Key == "" {
		s.ConnectionManager = def.ConnectionManager
		if s.EnvVars.Key == "" || s.EnvVars.Key == "e" {
			s.EnvVars = def.EnvVars
		}
	} else if s.EnvVars.Key == "" {
		s.EnvVars = def.EnvVars
	}
	if s.SystemSettings.Key == "" {
		s.SystemSettings = def.SystemSettings
	}
	if s.RefreshConfig.Key == "" {
		s.RefreshConfig = def.RefreshConfig
	}
	if s.Find.Key == "" {
		s.Find = def.Find
	}
	if s.Copy.Key == "" {
		s.Copy = def.Copy
	}
	if s.Paste.Key == "" {
		s.Paste = def.Paste
	}
	if s.ClearOutput.Key == "" {
		s.ClearOutput = def.ClearOutput
	}
	if s.CommandPalette.Key == "" {
		s.CommandPalette = def.CommandPalette
	}
	// 废弃 Ctrl+J 快速切换：清空旧绑定，避免残留
	s.QuickSwitcher = ShortcutBinding{}
	if s.PaneZoom.Key == "" {
		s.PaneZoom = def.PaneZoom
	}
	if s.NextTab.Key == "" {
		s.NextTab = def.NextTab
	}
	if s.PrevTab.Key == "" {
		s.PrevTab = def.PrevTab
	}
	if s.CloseTab.Key == "" {
		s.CloseTab = def.CloseTab
	}
	if s.ToggleBroadcast.Key == "" {
		s.ToggleBroadcast = def.ToggleBroadcast
	}
	if s.OpenSftp.Key == "" {
		s.OpenSftp = def.OpenSftp
	}
	if s.OpenLocalShell.Key == "" {
		s.OpenLocalShell = def.OpenLocalShell
	}
	if s.SplitFocusLeft.Key == "" {
		s.SplitFocusLeft = def.SplitFocusLeft
	}
	if s.SplitFocusRight.Key == "" {
		s.SplitFocusRight = def.SplitFocusRight
	}
	if s.SplitFocusUp.Key == "" {
		s.SplitFocusUp = def.SplitFocusUp
	}
	if s.SplitFocusDown.Key == "" {
		s.SplitFocusDown = def.SplitFocusDown
	}
}

// LoadShortcutSettings 从 ~/.flashdock/app_data.json 的 shortcuts 加载
func LoadShortcutSettings() (ShortcutSettings, error) {
	d, err := loadAppDataSection()
	if err != nil {
		return DefaultShortcutSettings(), err
	}
	s := d.Shortcuts
	fillShortcutDefaults(&s)
	if migrateKeyMapsIntoSnippets(&s) {
		_ = SaveShortcutSettings(s)
	}
	return s, nil
}

// migrateKeyMapsIntoSnippets 将旧版 keymaps.json 并入命令片段（一次性）
func migrateKeyMapsIntoSnippets(s *ShortcutSettings) bool {
	if s == nil {
		return false
	}
	km, err := LoadKeyMapSettings()
	if err != nil || len(km.Entries) == 0 {
		return false
	}

	existingIDs := make(map[string]struct{}, len(s.Snippets))
	for _, sn := range s.Snippets {
		if sn.ID != "" {
			existingIDs[sn.ID] = struct{}{}
		}
	}

	changed := false
	for _, e := range km.Entries {
		id := e.ID
		if id == "" {
			id = fmt.Sprintf("km-migrated-%d", len(s.Snippets)+1)
		}
		if _, ok := existingIDs[id]; ok {
			continue
		}
		cmd := e.SendString
		execute := false
		if strings.HasSuffix(cmd, `\n`) {
			cmd = strings.TrimSuffix(cmd, `\n`)
			execute = true
		} else if strings.HasSuffix(cmd, "\n") {
			cmd = strings.TrimSuffix(cmd, "\n")
			execute = true
		}
		name := strings.TrimSpace(e.Name)
		if name == "" {
			name = "按键映射"
		}
		binding := e.Binding
		sn := ShellSnippet{
			ID:      id,
			Name:    name,
			Command: cmd,
			Scope:   "global",
			Execute: execute,
		}
		if binding.Key != "" {
			b := binding
			sn.Binding = &b
		}
		s.Snippets = append(s.Snippets, sn)
		existingIDs[id] = struct{}{}
		changed = true
	}
	if !changed {
		return false
	}
	_ = SaveKeyMapSettings(DefaultKeyMapSettings())
	return true
}

// SaveShortcutSettings 保存到 ~/.flashdock/app_data.json 的 shortcuts
func SaveShortcutSettings(s ShortcutSettings) error {
	fillShortcutDefaults(&s)
	return updateAppData(func(d *AppDataFile) {
		d.Shortcuts = s
	})
}
